client/packages/electron/src/keyboardScanner/keyboardScanner.ts

```diff
@@ -0,0 +1,231 @@
+    // and at least minBarcodeLength long
+    this.interval = setInterval(() => {
+      this.processInputs();
+    }, PROCESSOR_INTERVAL);
+  }
+
+  stop() {
+    this.scanning = false;
+    if (this.interval) {
+      clearInterval(this.interval);
+      this.interval = undefined;
+    }
+  }
+
+  async processInputs() {
+    if (this.lockBuffer) return;
+    try {
+      this.lockBuffer = true;
+
+      let checkSequence: KeyEvent[] = [];
+      let passThrough: KeyEvent[] = [];
+      let barcode: KeyEvent[] = [];
+
+      this.buffer.forEach((current, index) => {
+        // Get ms for next input event or now
+        const nextMs = this.buffer[index + 1]?.ms ?? Date.now();
+        let diff = nextMs - current.ms;
+
+        checkSequence.push(current);
+        // Sequence is still ongoing
+        // ms diff between every event is sequence is below  maxMSBetweenScannerKeyInputs
+        if (diff < this.settings.maxMSBetweenScannerKeyInputs) return;
+        // If exceeding maxMSBetweenScannerKeyInputs and sequence length
+        // is not long enough to be barcode, pass all input through
+        if (checkSequence.length < this.settings.minBarcodeLength) {
+          passThrough = [...passThrough, ...checkSequence];
+          checkSequence = [];
+        } else {
+          // If exceeding maxMSBetweenScannerKeyInputs but sequnce length
+          // is long enought to be barcode, record as barcode;
+          // Very unlikely that two barcodes will be scanned within PROCESSOR_INTERVAL
+          barcode = checkSequence;
+          checkSequence = [];
+        }
+      });
+
+      // Remember left over for next check
+      this.buffer = checkSequence;
+      if (barcode.length != 0) {
+        console.log(JSON.stringify(barcode, null, ' '));
+        this.sendBarcode(barcode);
+      }
+
+      if (passThrough.length != 0) {
+        // Should allow pass through input events, even if bufferIsLocked
+        this.allowEvents = true;
+
+        passThrough.forEach(({ input: { key } }) => {
+          this.sendKey(key);
+        });
+
+        // Delay slightly to allow above events to go through, before unlocking buffer
+        await new Promise(resolve => setTimeout(resolve, 10));
+        this.allowEvents = false;
+      }
+    } catch (e) {
+      console.log(e);
+    } finally {
+      // Make sure lockBuffer is always unlocked
+      this.lockBuffer = false;
+    }
+  }
+
+  bindInputEvent() {
+    // Bind keyboard capture on startup (this.scanning is false by default)
+    this.window.webContents.on('before-input-event', (event, input) => {
+      if (!this.scanning) return;
+      if (keysToPassThrough.includes(input.key)) return;
+
+      while (this.lockBuffer) {
+        // Buffer will be locked when processing input, and when input is being processed and key event
+        // is identified as normal keyboard event it will be triggered from processor, this event should be
+        // passed through (no preventDefault)
```

Comment:
I'm finding the comment hard to follow. Some formatting might help - also unsure of what 'it' is and the 'event'
```   // Buffer will be locked when processing input
        // When input is being processed and key event is identified as normal keyboard event
        // it [what does 'it' mean here?] will be triggered from processor
        // this event [which event are we talking about?] should be passed through (no preventDefault) [hm?]
```

Reply:
Hmm, yeah that looks very hard to read, I changed based on questions in the example comment: 4a1cac38cdfbd0f43d28b2c916ba29d9159866a6, is this ok ?

Reply:
sweet - I understand now! 😄